---
title: "GWAS basic models"
author: "Filippo Biscarini"
date: "October 26, 2019"
output:
  ioslides_presentation:
      css: styles.css
  slidy_presentation: default
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
library("knitr")
library("qqman")
library("sommer")
library("rrBLUP")
library("ggplot2")
library("tidyverse")
library("factoextra")
library("data.table")

knitr::opts_chunk$set(echo = FALSE)
```

## GWAS: a one-SNP model

For a quantitative trait (continuous target variable), the basic model
is a **linear regression model** of the following form:

$$
y_{ij}=\mu+SNP_j+e_{ij}
$$

where $y_{ij}$ is the phenotypic observation for sample *i* with SNP
genotype *j*, $SNP_j$ is the effect of SNP genotype *j* (e.g. AA, AB,
BB), and $e_{ij}$ are the residuals of the model ($\mu$ is the
intercept).

### Loading the data

We use a dog dataset: *Nova Scotia Duck Tolling Retriever* (NSDTR), with
**125 dogs** on a genetic defect, *cleft palate*: **112 controls**, **13
cases**

The original data can be downloaded from here:
<https://datadryad.org/stash/dataset/doi:10.5061/dryad.j8r8q> (for this
practical, the data are already provided).

!! WARNING: we use here a binary trait as it were a continuous trait:
this is fine for the purpose of this practical, however keep in mind
that with binary variables logistic regression should be used in stead
of linear regression !!

Let's read in the data:

-   phenotypes: dog, family, status (case or control)
-   genotypes: dog, SNP genotypes (0/1/2)

```{r data_sheep, echo=TRUE}
phenotypes <- read.table("../data/dogs_phenotypes.txt", header = TRUE)
genotypes <- read.table(
  "../data/dogs_imputed.raw", 
  header = TRUE)
genotypes <- genotypes[,-c(1,3,4,5,6)]

## here we take just one SNP, to illustrate the basic GWAS model
df <- merge(phenotypes, genotypes[,c(1,25)], by.x="id", by.y="IID")
names(df) <- c("id","family","disease","snp")
```

```{r}
head(df)
```

### Modelling number of copies of the minor allele

We can model SNP genotypes as the number of copies of the minor allele:
0, 1, 2 (numeric variable). We apply the usual `R` syntax for a simple
**linear regression model**:

```{r model1, echo=TRUE}
fit <- lm(disease ~ snp, data = df)
```

### Results - summary(fit)

```{r echo=FALSE}
g <- summary(fit)
print(g)
```

### ANOVA table

```{r echo=TRUE}
anova(fit)
```

### Add population structure

The dogs belongs all to the same breed: there's no easy way in this case
to account for population structure by simply adding the
`breed variable` in the model. We can however extract the **principal
components of the genetic data** and use these as variables in the GWAS
model to account for population structure.

#### PC's

-   `princomp()`: **spectral decomposition** which examines the
    covariances / correlations between variables
-   `prcomp()`, `FactoMineR::PCA()`: **singular value decomposition**
    (**SVD**) which examines the covariances / correlations between
    individuals

```{r echo=TRUE}
X <- genotypes[,-1]
Xscaled = scale(X, center = TRUE, scale = TRUE)

## uncomment if you want to recalculate PCs from SNP data
## (this may take a while)
res <- prcomp(X, rank. = 3)

## we can load precomputed PCs from the RData object in /data
# load("../data/pca.RData")
# names(res)

npcs = length(res$sdev)
print(paste("QUESTION: The total number of PCs is", npcs, ": can you guess why?"))
```

#### Eigenvalues

```{r echo=TRUE}
eig <- (res$sdev)^2
variance <- eig*100/sum(eig)
```

#### Barplot of eigenvalues

```{r echo=TRUE}
barplot(variance, col="red", xlab = "Principal Components", 
        ylab = "% variance")
sum(variance[c(1:3)])
```

```{r echo=TRUE}
#library("factoextra")
fviz_eig(res, ncp = 25)
```

#### Use principal components to model population structure

```{r}
head(res$x[,c(1,2,3)])

## we use the first 3 PCs
df <- cbind.data.frame(df,res$x[,c(1:3)])
```

```{r echo=TRUE}
fit <- lm(
  disease ~ snp + PC1 + PC2 + PC3, 
  data = df
)
```

#### Results - summary(fit) {.smaller}

```{r echo=FALSE}
g <- summary(fit)
print(g)
```

## Multi-SNP model

Up to here we have applied our GWAS model to **only one SNP**; however,
**typical GWAS projects include thousands of SNPs**. Let's see how to do
GWAS with our entire dataset.

We have: - 125 dogs - one phenotype (disease): 112 controls, 13 cases -
genotypes at $10\,899$ SNP loci

### SNP genotypes

```{r}
## READING DATA
print("now reading in the data ...")

### genotypes (we already read this file at the beginning of this notebook)
print(paste(nrow(genotypes),"records read from the genotype file",sep=" "))

## read the SNP map file
SNP_INFO <- fread("../data/dogs_imputed.map")
names(SNP_INFO) <- c("Chr","SNP","cM","Pos")
SNP_INFO$cM <- NULL
```

```{r}
X <- as.matrix(genotypes[,-1]) ## first column: dog ID
colnames(X) <- gsub("\\_[A-Z]{1}$","",colnames(X))
rownames(X) <- genotypes$IID

SNP_INFO <- bind_cols(SNP_INFO,as.data.frame(t(X)))

print(paste(nrow(SNP_INFO),"SNPs read from the map file",sep=" "))
```

### Phenotypes

```{r}
## sanity check
phenotypes <- phenotypes[phenotypes$id %in% genotypes$IID,]
phenotypes <- phenotypes[,c(1,3)]
```

```{r}
head(phenotypes)
```

### Kinship matrix

We saw that one option to account for population structure in the GWAS
models is to use principal components extracted from the SNP genotype
data.

Better still, we can use the kinship matrix (that we introduced in a
previous module): in this way, we can use the entire genomic information
available, rather than just that encapsulated in the first principal
components.

Additionally, we can use the genomic information in the form of a
covariance matrix, rather than as additive systematic effects.

```{r}
K <- sommer::A.mat(X)
colnames(K) <- genotypes$IID
rownames(K) <- genotypes$IID

vec <- colnames(K) %in% phenotypes$id
K <- K[vec,vec]
```

```{r}
heatmap(K,col=rev(heat.colors(75)))
```

### Run the multi-gwas model

```{r}
model1_x <- rrBLUP::GWAS(
   pheno = phenotypes,
   geno = SNP_INFO,
   # fixed = "population",
   K = K,
   plot = FALSE
)
```

```{r}
# model1_x <- sommer::GWAS(
#  fixed = phenotype~1, 
#  random = ~vsr(id, Gu=K),
#  rcov = ~units,
#  data = phenotypes,
#  M = X,
#  gTerm = "u:id",
#  verbose = TRUE
#)
```

### RESULTS

File with results from GWAS:

```{r}
print("writing out results and figures ...")
names(model1_x)[length(model1_x)] <- "disease"
gwasResults <- model1_x[,c("SNP","Chr","Pos","disease")]
names(gwasResults) <- c("SNP","CHR","BP","P")
head(gwasResults) |> kable()
```

#### Manhattan plot

```{r}
qqman::manhattan(gwasResults, suggestiveline = TRUE, col = c("red","blue"),
          genomewideline = FALSE, logp = FALSE)

```

#### QQ-plot

```{r}
# convert -log(p) back to p-values
p <- 10^((-gwasResults$P))

## rename P to log_p (as it is) and add the column with p-values
names(gwasResults)[4] <- "log_p"
gwasResults$P <- p

## qq-plot
qqman::qq(p)
```
